loader: Do not remove duplicated layers

RyuzakiKK opened the pull request:
Linux can support multiple ABIs, but Vulkan Layer manifest does not
allow to specify different library paths, based on the ABI.

As a solution, for ICDs, we can simply create multiple manifests, one
per ABI, and the Loader will try them one by one until it finds the
library that is compatible with the executable class.

Instead, for Vulkan Layers, this method doesn't work because the Loader
will discard the manifests that have a duplicated layer name.

To add support for multiple ABIs to Vulkan Layers, and to make the
behavior similar to the ICDs, with this commit we remove the duplicated
layer name check.

Fixes: #155

---

Please check @smcv's comment for more information about this issue and the possible solutions https://github.com/KhronosGroup/Vulkan-Loader/issues/155#issuecomment-736630098

lenny-lunarg commented:
Have you checked what happens when there are duplicated layers and both can be loaded by the loader? It looks to me like both will get loaded, which isn't what we want. I'm broadly fine with the idea of of just stopping the de-duplication when the layers are found and using the same mechanism that people have used for ICDs. The problem then comes when you have multiple layers with the same name and both can be loaded (which is why the de-duplication logic exists now).

My thinking here is that I'd probably be willing to take this change, but only after we've added some de-duplication logic back at the point where we actually `dlopen` and load the layers. This should allow us to build a list with both versions of a layer, and then only actually load the first layer with a given name that can actually be loaded.

smcv commented:
> Have you checked what happens when there are duplicated layers and both can be loaded by the loader?

Is the situation you're concerned about something like this? You have a new version of vkBasalt in `~/.local` and an old version in `/usr`; with the current loader you'd only load the version from `~/.local` which is presumably the one you wanted; but with this PR, you'd load both and they'll probably fight?

smcv commented:
Continuing from that, let's suppose you have a new 64-bit-only version in `~/.local`, and old versions for both 32- and 64-bit in `/usr`. I think the result we want is probably that 64-bit executables load the new version from `~/.local` and ignore the older version in `/usr`, while 32-bit executables load the older version from `/usr` because it's the best available for that ABI? Do you agree?

RyuzakiKK commented:
@lenny-lunarg Thanks for your feedback.
To avoid dlopening layers multiple times I changed a bit the de-duplication logic with the latest commit that I pushed.

As previously, I removed the `loaderListHasLayerProperty()` check, so that we add to the layer list all the layers that we find, even if duplicated.

But then, before actually opening the layer, I added a check `loaderNamesArrayHasLayerProperty()` to skip layers that have the same name as the layers that we already successfully activated.

In this way if we have more than one manifest with the same name, but for different ABIs, we will try to open them one by one until we find the loadable one.
Instead if we have multiple manifests with the same name and even the same library_path  (or different library_path but all loadable), as soon as we are able to activate one of them, all the following will be discarded.

smcv commented on loader/loader.c:
```diff
@@ -6043,6 +6031,8 @@ VkResult loader_create_instance_chain(const VkInstanceCreateInfo *pCreateInfo, c
 
             chain_info.u.pLayerInfo = &layer_instance_link_info[activated_layers];
 
+            activated_layer_names[activated_layers] = strdup(layer_prop->info.layerName);
```
Looks like this is leaked? I think you'll either need to "borrow" the string (if the lifetime of `layer_prop->info.layerName` is guaranteed to be longer than `activated_layer_names`), or free the copied strings stored in `activated_layer_names` before it goes out of scope (otherwise).

If you're using `strdup()`, note that it can fail and return `NULL`.

RyuzakiKK replied:
Ops, your are right. I guess I misinterpreted `alloca()` and forgot to free the actual elements.

I removed the `strdup()` because `layer_prop` is just a pointer to `dev->expanded_activated_layer_list.list[i]`, that does not change while we are in this "for" loop. So it is guaranteed to last longer than `activated_layer_names`.

smcv commented on loader/loader.c:
```diff
@@ -6246,6 +6249,8 @@ VkResult loader_create_device_chain(const VkPhysicalDevice pd, const VkDeviceCre
             nextGIPA = fpGIPA;
             nextGDPA = fpGDPA;
 
+            activated_layer_names[activated_layers] = strdup(layer_prop->info.layerName);
```
Same here.

smcv commented:
This updated version looks good to me, FWIW.

Plagman commented:
@lenny-lunarg can you help review this? This would be great to get in soon for Proton, assuming it looks good.

lenny-lunarg commented:
@Plagman, yeah, I'll get to it this afternoon. Broadly I can say that it looks good, but I want to take the time to do a little testing and make sure it works the way I think it does.

charles-lunarg commented:
What happens if the oolder loader is on a users system but a newer layer is installed, specifically one where the layer has 2 manifests, one for each platform & they share the same name? Would the loader remove one of the layers at random?
While allowing duplicate names (until dlopen is called) works going forward, any changes to the layer interface also should work with older versions, unless a breakage is intentional.

RyuzakiKK commented:
> What happens if the older loader is on a users system but a newer layer is installed, specifically one where the layer has 2 manifests, one for each platform & they share the same name? Would the loader remove one of the layers at random?

Yes, that's what will happen.

> While allowing duplicate names (until dlopen is called) works going forward, any changes to the layer interface also should work with older versions, unless a breakage is intentional.

I see your point, but I'm not quite sure how the best way forward would be. Right now, if you have multiple layers with the same name, the loader just picks at random and there is no way around that.

Probably we could use another layer manifest version to signal the fact that a layer is requiring a new loader version that supports defining multiple layers with the same name?

@charles-lunarg What do you think? Do you have something else in mind?

charles-lunarg commented:
I'm not sure the layer interface version will help anything here, as older loaders wont ignore layers with a newer version, just emit a warning about having an unsupported layer manifest version.

I am thinking that it would be feasible to add a 'platform' field to the json manifest, which indicates the platform. This would make it so that instead of allowing duplicates in the list itself, the loader can cull manifests which do not match the current platform. 
```json
{
    "platform": "64"
}
```
It does allow for layers with the same name, because the layers that would have the same name are removed from consideration before ever being added to the manifest list. One benefit is that the loader logic would be simplified since it can keep the deduplication logic in place and not delay it until dlopen is called, so it should be a bit faster as it wont do any extraneous calls to dlopen.
However, it still suffers from the same issue I listed above, as a layer that has two manifests with the same name and this new field would have the loader ignore the field and pick on of the layers at random. The underlying issue is that any solution which allows for layer manifest to have the same file name will always be incompatible with older loader versions.

RyuzakiKK commented:
> I'm not sure the layer interface version will help anything here, as older loaders wont ignore layers with a newer version, just emit a warning about having an unsupported layer manifest version.

Ah okay, I thought they were discarded.

> I am thinking that it would be feasible to add a 'platform' field to the json manifest, which indicates the platform. This would make it so that instead of allowing duplicates in the list itself, the loader can cull manifests which do not match the current platform.
> 
> ```json
> {
>     "platform": "64"
> }
> ```
> 
> It does allow for layers with the same name, because the layers that would have the same name are removed from consideration before ever being added to the manifest list. One benefit is that the loader logic would be simplified since it can keep the deduplication logic in place and not delay it until dlopen is called, so it should be a bit faster as it wont do any extraneous calls to dlopen.

Please check this @smcv comment about the reasoning why a platform specific field should be avoided if possible https://github.com/KhronosGroup/Vulkan-Loader/issues/155#issuecomment-736630098

charles-lunarg commented:
From my understanding of the ELF file format, the 5th byte in the header of the file signifies whether the ELF file is 32 or 64 bit. That would be which 'platform' that would be considered, not ABI. ELF calls this 'class' so maybe platform is the wrong word for the situation.
My thought is that it would only have 32 and 64 as valid options, nothing more. It is indeed a can of worms to have every known ABI/platform as a supported option. This class would also be an optional field, as that helps with deployment issues but also in cases where such a description isn't helpful for whatever reason. Such a field also wouldn't interfere with allowing duplicate layer names, as that could be necessary if 32 and 64 isn't sufficient. The goal of this field is to allow the loader to proactively remove layers that will fail `dlopen` with the wrong clas, saving time. Calling to dlopen to determine the class should be a "last resort" and so a proactive approach is better than reactive.

smcv commented:
Oh, if you only mean optimizing this pseudocode:

```
if (dlopen or platform equivalent succeeds) {
    add to list
}
else {
    log a warning, ignore
}
```

into this

```
if (JSON says it's 32-bit && sizeof (void *) != 4) {
    ignore, not for our word size
}
else if (JSON says it's 64-bit && sizeof (void *) != 8) {
    ignore, not for our word size
}
else if (dlopen or platform equivalent succeeds) {
    add to list
}
else {
    log a warning, ignore
}
```

then that could make sense. I think the behaviour in the absence of a word-size marker would still need to be reasonable - it would be more like an optimization than a bug fix, and I think this PR would still be desirable, although it would also have the beneficial side-effect of avoiding #262 on x86.

If that's the goal, then calling this marker "word size" rather than "platform" is probably the clearest way to think about it.

It isn't entirely trivial to find out your own ELF class without using relatively obscure corners of glibc (I think you have to retrieve the link map and inspect one of the objects), and in general you can't answer the question "if I dlopen()'d the library given by this arbitrary string, what filename would actually get opened?" without literally doing the dlopen operation, at which point it's too late to be optimizing away a dlopen that is going to fail by looking at the ELF header. However, it *is* fairly straightforward to find out (even at compile time, and with some tricks, even when cross-compiling) how large a pointer is on your platform, and this is the sort of thing build systems need to be able to do anyway.

The reason I'm being careful to distinguish between size of pointer and ELF class is that I'm not sure whether they are *guaranteed* to match up on weird architectures like x32 and aarch64-ilp32 - so it seems best to choose one and stick to it, and using the size of a pointer is going to be a lot easier.

charles-lunarg commented:
Yes, that is exactly what I was intending, a way to make it possible to bypass calling dlopen and simply fallback to that when it isn't present. 

You bring up very important points, ones that need to be nailed down before any such solution is implemented, namely if the pointer size can be guaranteed to match the elf class and how to figure out at compile time which size to use. The latter should be straightforward, as even vulkan_core.h has this in the NON_DISPATCHABLE_HANDLE macro. The former I wouldn't be confident in being guaranteed, but don't think it is extremely likely, and if it does occur, should be straightforward to diagnose.

On the topic of ICD's, I am not confident this solution is desired, as there was a notion that driver manifests aren't necessary for ICD's. I personally would like to add this, but would need more confidence it'd be used and won't cause issues with old loaders or other drivers in the wild before committing.

Lastly, I do think this PR is still useful, as it brings back the same behavior between layer and ICD manifest searching. It also makes it 'simpler' if you were writing a layer for both 32 and 64 bit.

lenny-lunarg reviewed (approved):
My apologies for not doing this sooner. The PR looks good to me.
